### PR TITLE
fix Matrix*list

### LINKED ISSUE #13774: Odd behavior with Matrix*list
```
>>> M = Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
>>> v = [1, 1, 1]
>>> M*v
Matrix([
[                                                      [1, 1, 1],                                                       [1, 1, 1, 1, 1, 1],                                                       [1, 1, 1, 1, 1, 1, 1, 1, 1]],
[                           [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],                            [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],                            [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]],
[[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]])
```

This is based on the example in the Matrix.dot docstring. The result should probably be an exception, or an automatic conversion of the list to a matrix. The matrix of lists is wrong, and I'm surprised it's even allowed.

### PR DESCRIPTION
**Pull request by vishalg2235**
Fixes #13774 
It raises TypeError for any `Matrix*Iterable`.

### CODE REVIEW

**Comment by vishalg2235** [comment]
ping @asmeurer

File: sympy/matrices/common.py [hunk 1]
```diff
@@ -2018,10 +2018,12 @@ def __mul__(self, other):
         if getattr(other, 'is_MatrixLike', False):
             return MatrixArithmetic._eval_matrix_mul(self, other)
 
-        try:
-            return self._eval_scalar_mul(other)
-        except TypeError:
-            pass
+        #if 'other' is not iterable then scalar multiplication.
+        if not isinstance(other,collections.Iterable):
```
Comment (siefkenj) [inline sympy/matrices/common.py]: `isinstance(other, collections.Iterable)` with the space :-)

File: sympy/matrices/common.py [hunk 2]
```diff
@@ -2018,10 +2018,12 @@ def __mul__(self, other):
         if getattr(other, 'is_MatrixLike', False):
             return MatrixArithmetic._eval_matrix_mul(self, other)
 
-        try:
-            return self._eval_scalar_mul(other)
-        except TypeError:
-            pass
+        #if 'other' is not iterable then scalar multiplication.
```
Comment (asmeurer) [inline sympy/matrices/common.py]: Should also put a space after the #

File: sympy/matrices/tests/test_commonmatrix.py [hunk 3]
```diff
@@ -1340,3 +1340,8 @@ def test_limit():
     x, y = symbols('x y')
     m = CalculusOnlyMatrix(2, 1, [1/x, y])
     assert m.limit(x, 5) == Matrix(2, 1, [S(1)/5, y])
+
+def test_issue_13774():
+    M = Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
+    v = [1,1,1]
+    raises(TypeError, lambda: M*v)
```
Comment (asmeurer) [inline sympy/matrices/tests/test_commonmatrix.py]: Also add a test for v*M

**Comment by asmeurer** [comment]
Looks good other than those comments.

File: sympy/matrices/tests/test_commonmatrix.py (hunk 3 above)
Comment (vishalg2235) [reply sympy/matrices/tests/test_commonmatrix.py]: for v*M again same thing is happening which was previously happening for M*v. Now it seems scalar multiplication error. Where in code should I fix it ?

Comment (vishalg2235) [reply sympy/matrices/tests/test_commonmatrix.py]: From where in code this is happening ?
```
>>> 8*A
Matrix([
[ 8, 16],
[24, 32]])
>>> [1,2,'r']*A
Matrix([
[                  [1, 2, r],                   [1, 2, r, 1, 2, r]],
[[1, 2, r, 1, 2, r, 1, 2, r], [1, 2, r, 1, 2, r, 1, 2, r, 1, 2, r]]])
```

Comment (vishalg2235) [reply sympy/matrices/tests/test_commonmatrix.py]: Ping @asmeurer

**Comment by vishalg2235** [comment]
`v*M` is also fixed and test case added.

**Review by siefkenj — APPROVED** [review]
Looks good

File: sympy/matrices/tests/test_commonmatrix.py (hunk 3 above)
Comment (asmeurer) [reply sympy/matrices/tests/test_commonmatrix.py]: Please also add a test for `v*M`.

**Comment by asmeurer** [comment]
This just needs a test for `v*M`. Other than that it looks good.

File: sympy/matrices/tests/test_commonmatrix.py [hunk 4]
```diff
@@ -1340,3 +1340,9 @@ def test_limit():
...
+    raises(TypeError, lambda: v*M)
```
Comment (vishalg2235) [inline sympy/matrices/tests/test_commonmatrix.py]: @asmeurer here is the test for `v*M`.

File: sympy/matrices/common.py [hunk 5]
```diff
@@ -2018,10 +2018,12 @@ def __mul__(self, other):
...
-        except TypeError:
-            pass
+        # if 'other' is not iterable then scalar multiplication.
+        if not isinstance(other, collections.Iterable):
+            try:
+                return self._eval_scalar_mul(other)
+            except TypeError:
+                pass
 
         raise TypeError('Cannot multiply %s and %s' % (type(self), type(other)))
```
Comment (ylemkimon) [inline sympy/matrices/common.py]: I think it's better to return `NotImplemented`, as Python automatically constructs and raises `TypeError` if `NotImplemented` is returned by overridden Python operations.

Comment (vishalg2235) [reply sympy/matrices/common.py]: Yes. `NotImplemented` is better. But I wonder in which case it will not be implemented ?

Comment (ylemkimon) [reply sympy/matrices/common.py]: As you said, in cases of `Matrix*Iterable`.

Comment (vishalg2235) [reply sympy/matrices/common.py]: but `iterable` will be omitted after that `if` statement.

Comment (vishalg2235) [reply sympy/matrices/common.py]: oh sorry I wanted to ask about line 2025 `except`. In which case code will raise `TypeError`.

Comment (ylemkimon) [reply sympy/matrices/common.py]: Multiplication that makes no sense like `None*Matrix` would raise it.

Comment (vishalg2235) [reply sympy/matrices/common.py]: Ok Thanks.

**Comment by ylemkimon** [comment]
I meant 'return NotImplemented', not 'raise NotImplementedError'. See https://docs.python.org/3/library/constants.html#NotImplemented.

**Comment by vishalg2235** [comment]
ok !!

**Comment by ylemkimon** [comment]
Thank you for your contribution!

**Pull request merged**